src/DataUtils.cpp

```diff
@@ -0,0 +1,211 @@
+#include "NAS2D/DataUtils.h"
+
+#include "NAS2D/Xml/XmlAttribute.h"
+#include "NAS2D/Xml/XmlElement.h"
+
+#include <algorithm>
+#include <sstream>
+
+#if defined(_WIN32)
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#endif
+
+void NAS2D::Xml::validateXMLElement(
+	const NAS2D::Xml::XmlElement& element,
+	const std::string& name,
+	const std::string& requiredChildElements,
+	const std::string& requiredAttributes,
+	const std::string& optionalChildElements,
+	const std::string& optionalAttributes)
+{
+	if (name.empty())
+	{
+		// TODO: Make this a function with [[noreturn]] attribute
+#if defined(_WIN32)
+		std::ostringstream err_ss;
+		err_ss << "Xml Validation failed. Element name is required.";
+		::MessageBoxA(nullptr, err_ss.str().c_str(), "Fatal Error", MB_ICONERROR | MB_OK);
+#else
+		// TODO: Add non-windows version.
+#endif
+		exit(0);
+	}
+
+	const std::string elemName = element.value();
+	if (elemName != name)
+	{
+		// TODO: Make this a function with [[noreturn]] attribute
+#if defined(_WIN32)
+		std::ostringstream err_ss;
+		err_ss << "Xml Validation failed. Element name " << elemName << " does not match valid name " << name;
+		::MessageBoxA(nullptr, err_ss.str().c_str(), "Fatal Error", MB_ICONERROR | MB_OK);
+#else
+		// TODO: Add non-windows version.
+#endif
+		exit(0);
+	}
+
+	// Get list of required/optional attributes/children
+	// Sort
+	// Remove duplicates
+	// Rational for not using std:set:
+	// Profiled code takes average of 10 microseconds to complete.
+	auto requiredChildNames = NAS2D::split(requiredChildElements);
```

Comment:
I'm actually wondering why this wasn't passed in as an array (`std::vector<std::string>`).

Reply:
So you don't have to build a a vector beforehand and the list can just be a CSV list C-string.

See: https://github.com/lairworks/nas2d-core/issues/434#issuecomment-583219252

Reply:
I was just thinking. On another project, I recently discovered that `std::string_view` is `constexpr` compatible, even though `std::string` is not. In that project I was able to create an array of string_views at compile time. That might be one way to set a static list of parameters without incurring runtime cost.

I may have been using `std::array` rather than `std::vector` though, so that may impact the usefulness of the technique.

Reply:
You might get away with `std::string_view` parameters into the validation function but they are still being split up into separate strings. So at most you're saving 5 string constructions where SSO in most typical cases would handle 3 of them.